taichi/ir/control_flow_graph.cpp

```diff
@@ -863,4 +864,104 @@ std::unordered_set<SNode *> ControlFlowGraph::gather_loaded_snodes() {
   return snodes;
 }
 
+void ControlFlowGraph::determine_ad_stack_size(int max_ad_stack_size) {
+  const int num_nodes = size();
+  std::unordered_map<AdStackAllocaStmt *, int> max_size;
+  std::vector<std::unordered_map<AdStackAllocaStmt *, int>>
+      max_size_at_node_begin(num_nodes);
+  std::vector<std::unordered_map<AdStackAllocaStmt *, int>> max_increased_size(
+      num_nodes);
+  std::vector<std::unordered_map<AdStackAllocaStmt *, int>> increased_size(
+      num_nodes);
+  std::queue<int> to_visit;
+  std::vector<bool> in_queue(num_nodes);
+  std::unordered_map<CFGNode *, int> node_ids;
+  std::unordered_set<AdStackAllocaStmt *> oversized_stacks;
+
+  for (int i = 0; i < num_nodes; i++)
+    node_ids[nodes[i].get()] = i;
+
+  for (int i = 0; i < num_nodes; i++) {
+    for (int j = nodes[i]->begin_location; j < nodes[i]->end_location; j++) {
+      Stmt *stmt = nodes[i]->block->statements[j].get();
+      if (auto *stack_push = stmt->cast<AdStackPushStmt>()) {
+        auto *stack = stack_push->stack->as<AdStackAllocaStmt>();
+        if (stack->max_size == 0 /*adaptive*/) {
+          increased_size[i][stack]++;
+          if (increased_size[i][stack] > max_increased_size[i][stack]) {
+            max_increased_size[i][stack] = increased_size[i][stack];
+          }
+        }
+      } else if (auto *stack_pop = stmt->cast<AdStackPopStmt>()) {
+        auto *stack = stack_pop->stack->as<AdStackAllocaStmt>();
+        if (stack->max_size == 0 /*adaptive*/) {
+          increased_size[i][stack]--;
+        }
+      }
+    }
+    to_visit.push(i);
+    in_queue[i] = true;
+  }
```

Comment:
Comment something like "For each basic block we compute the increase of stack size. This is a pre-processing step for the next maximum stack determining algorithm..."